### PR TITLE
Add voters in batches

### LINKED ISSUE #155: Cannot add 5'000 scipers for an election
When I add 5'000 scipers as voters, I get : 

image

The election is created but I cannot do anything : 

image

And I don't see anymore the button "Create form" (as I am not anymore an admin). But if I refresh the page, I can get back the button "Create form" ...

If I put my sciper number at the top of the list of the scipers number, I can vote even with the error message above. But if my sciper number is at the bottom of the list of the scipers number, I am not allowed to vote ... 

> [!NOTE]
> It seems I can add batches of 300 to 500 (max) sciper numbers. 
> It would be useful to know how many sciper numbers have been added. The window of "Adding voters" keeps growing too much with a large number of scipers. A vertical scroll woulb be enough.

### PR DESCRIPTION
**Pull request by PascalinDe**
This PR adds voters in batches instead of one-by-one, allowing to add > 5000 voters at a time. Closes #155 .

### CODE REVIEW

File: web/backend/src/controllers/users.ts [hunk 1]
```diff
@@ -34,14 +34,28 @@ usersRouter.post('/add_role', (req, res, next) => {
...
 
-  addPolicy(req.body.userId, req.body.subject, req.body.permission)
-    .then(() => {
-      res.set(200).send();
-      next();
-    })
-    .catch((e) => {
-      res.status(400).send(`Error while adding to roles: ${e}`);
-    });
+  if (Object.hasOwn(req.body, 'userId')) {
```
Comment (ineiti) [inline web/backend/src/controllers/users.ts:37]: For next time: I think `'userId' in req.body` would be clearer to me. Same below.

File: web/backend/src/controllers/users.ts [hunk 2]
```diff
@@ -34,14 +34,28 @@ usersRouter.post('/add_role', (req, res, next) => {
...
+    addPolicy(req.body.userId, req.body.subject, req.body.permission)
+      .then(() => {
+        res.set(200).send();
+        next();
+      })
+      .catch((e) => {
+        res.status(400).send(`Error while adding to roles: ${e}`);
+      });
```
Comment (ineiti) [inline web/backend/src/controllers/users.ts:45]: You have to be careful with the `then` construct here, as it will continue executing the code below, which might not be what you want. If you think this is a problem (I didn't look at it in detail), you should change it to the following. Same for the `then` just below.

```
  try {
    await addPolicy(req.body.userId, req.body.subject, req.body.permission);
    res.set(200).send();
    next();
  } catch(e) {
    res.status(400).send(`Error while adding to roles: ${e}`);
  };
```

**Review by ineiti — APPROVED** [review]
There is one part with the `then` where I'm not sure the code does what you want it to. Else it looks good to me.

File: web/backend/src/controllers/users.ts [hunk 3]
```diff
@@ -34,17 +34,29 @@ usersRouter.post('/add_role', (req, res, next) => {
...
-    })
-    .catch((e) => {
-      res.status(400).send(`Error while adding to roles: ${e}`);
-    });
-
-  // Call https://search-api.epfl.ch/api/ldap?q=228271, if the answer is
-  // empty then sciper unknown, otherwise add it in userDB
+  if ('userId' in req.body) {
+    try {
+      addPolicy(req.body.userId, req.body.subject, req.body.permission);
```
Comment (ineiti) [inline web/backend/src/controllers/users.ts:39]:
```suggestion
      await addPolicy(req.body.userId, req.body.subject, req.body.permission);
```

And you'll also need to add an `async` somewhere in the definition of the callback, IIRC, like that:

```
usersRouter.post('/add_role', async (req, res, next) => {
```

File: web/backend/src/controllers/users.ts [hunk 4]
```diff
@@ -34,17 +34,29 @@ usersRouter.post('/add_role', (req, res, next) => {
...
+    } catch (error) {
+      res.status(400).send(`Error while adding single user to roles: ${error}`);
+      return;
+    }
+    res.set(200).send();
+    next();
+  } else if ('userIds' in req.body) {
+    try {
+      addListPolicy(req.body.userIds, req.body.subject, req.body.permission);
```
Comment (ineiti) [inline web/backend/src/controllers/users.ts:48]:
```suggestion
      await addListPolicy(req.body.userIds, req.body.subject, req.body.permission);
```

**Review by ineiti — CHANGES_REQUESTED** [review]
Now it's important that you add the `await`, else the `catch` will never be called:

https://www.typescriptlang.org/play/?#code/IYZwngdgxgBAZgV2gFwJYHsIwBYFMA2+6AFAJQDeAUDDTFJiOvrgHREDmxARHoel6QDc1WsmwAndAHcYEXDICi4yeO7ZUA4QF9Ku0JFiIUGLAFtgqCGSo1k4sDBu0aAehcwA6qjHoEyGMBYAOTAUhbIQTi44rgANDBiuDDRKvCSpjBBvESRqCCyuABu0XTACOzYyCwiztkkQiJapchQ2DDEuKSONbT0EIzMLCnoqp3CNDp9A6wcahoNOpTmlmTCQA

File: web/frontend/src/pages/form/components/utils/useChangeAction.tsx [hunk 5]
```diff
@@ -314,35 +314,64 @@ const useChangeAction = (
...
+      setUserConfirmedAddVoters('');
+
+      for (const sciperStr of providedScipers) {
+        const sciper = parseInt(sciperStr, 10);
+        if (isNaN(sciper)) {
+          sciperErrs += t('sciperNaN', { sciperStr: sciperStr });
+        }
+        if (sciper < 100000 || sciper > 999999) {
+          sciperErrs += t('sciperOutOfRange', { sciper: sciper });
+        }
```
Comment (ineiti) [inline web/frontend/src/pages/form/components/utils/useChangeAction.tsx:329]: This should use `authManager.readSCIPER`

**Review by ineiti — CHANGES_REQUESTED** [review]

File: web/frontend/src/pages/form/components/utils/useChangeAction.tsx (hunk 5 above)
Comment (PascalinDe) [reply web/frontend/src/pages/form/components/utils/useChangeAction.tsx:329]: but that's defined in the backend

File: web/backend/src/controllers/users.ts (hunk 3 above)
Comment (PascalinDe) [reply web/backend/src/controllers/users.ts]: used `Promise.resolve` because I wasn't sure about the async in the callback, does that make sense?

File: web/backend/src/controllers/users.ts [hunk 6]
```diff
@@ -34,17 +41,31 @@ usersRouter.post('/add_role', (req, res, next) => {
...
-    .catch((e) => {
-      res.status(400).send(`Error while adding to roles: ${e}`);
-    });
-
-  // Call https://search-api.epfl.ch/api/ldap?q=228271, if the answer is
-  // empty then sciper unknown, otherwise add it in userDB
+  if ('userId' in req.body) {
+    try {
+      readSCIPER(req.body.userId);
+      Promise.resolve(addPolicy(req.body.userId, req.body.subject, req.body.permission));
```
Comment (ineiti) [inline web/backend/src/controllers/users.ts:47]: The `Promise.resolve` doesn't change anything - the promise will not be awaited for. Does the following PR break it?

https://github.com/c4dt/d-voting/pull/159

You can run [Await Playground](https://www.typescriptlang.org/play/?#code/IYZwngdgxgBAZgV2gFwJYHsIwBYFMA2+6AFAJQDeAUDDTFJiOvrgHREDmxARHoel6QDc1WsmwAndAHcYEXDICi4yeO7t06ACYAjMLgHCAvpROhIsRCgxYAtsFQQyVGsnFgYz2jQD03mAHVUMXQEZBhgLAByYCl7ZEicXHFcABoYMVwYJJV4SRsYSN4iBNQQWVwANyS6YAR2bGQWES8ABTzS1mTGfCriIpJSIRFDGuQobBhiXFIPZtp6CG7WbPRVaeEaYxoFpbZ0Th5UAxEYuPKZNvQbDuJicRmAXgA+GBBcZAAVVBtcEOQ7tIARgADKDBhs6AwmKwONwMsljsZKHYHGRhEA) and see that `goodbye` is never called.

**Review by ineiti — CHANGES_REQUESTED** [review]
I'm still surprised it actually works - according to my test in the typescript playground, the async should never be called.

File: web/backend/src/controllers/users.ts (hunk 6 above)
Comment (PascalinDe) [reply web/backend/src/controllers/users.ts:47]: I wasn't sure about the async in the callback :confused: I'll try with async/await

File: web/frontend/src/pages/form/components/utils/useChangeAction.tsx (hunk 5 above)
Comment (ineiti) [reply web/frontend/src/pages/form/components/utils/useChangeAction.tsx:329]: Oups...

File: web/backend/src/controllers/users.ts (hunk 6 above)
Comment (ineiti) [reply web/backend/src/controllers/users.ts]: Here is a thread from stackoverflow: https://stackoverflow.com/questions/44813401/passing-in-async-functions-to-node-js-express-js-router

TLDR: 
- express >= 4 handles async callbacks
- express >= 5 handles also errors w/o a try...catch

**Review by ineiti — APPROVED** [review]
